### PR TITLE
Enforce TRY linting with custom exceptions

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- enable TRY checks in ruff configuration
- replace generic RuntimeError strings with custom exceptions across client and server
- update tests to use new exception types

## Testing
- `make fmt`
- `make lint`
- `make test`
- `make typecheck`

------
https://chatgpt.com/codex/tasks/task_e_68a67ba7de64832290eeb23822c28904

## Summary by Sourcery

Enforce TRY linting and improve error handling by defining and using custom exception classes in place of generic errors across the codebase, and update tests to expect these new exception types.

Enhancements:
- Introduce custom exception classes in serena_tools, server, client, and test utilities for more precise error reporting
- Replace all generic RuntimeError, TypeError, ValueError, and TimeoutError raises with the corresponding custom exceptions

Build:
- Enable the TRY rule in the Ruff configuration for exception handling best practices

Tests:
- Update unit and feature tests to assert on the new custom exception types instead of generic ones

## Summary by CodeRabbit

- **New Features**
  - Clearer, more specific user-facing error types for daemon startup, onboarding, diagnostics, definition/reference lookup, and external tool integration.

- **Bug Fixes**
  - Improved daemon/socket readiness handling with clearer exit/reporting when connections or startup fail.

- **Tests**
  - Updated and added tests to cover new error signaling, client RPC connection/startup behavior, and Serena integration edge cases.

- **Chores**
  - Project metadata, packaging and dev-tool configs updated (author, Python 3.11 target, dev deps, lint/type settings).

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

```
Complex Conditional
weaverd/serena_tools.py: _import_serena_module
Suppress finding
What lead to degradation?

_import_serena_module has 1 complex conditionals with 2 branches, threshold = 2
Why does this problem occur?

A complex conditional is an expression inside a branch such as an if-statmeent which consists of multiple, logical operations. Example: if (x.started() && y.running()).Complex conditionals make the code even harder to read, and contribute to the Complex Method code smell. Encapsulate them.
How to fix it?

Apply the [DECOMPOSE CONDITIONAL](https://refactoring.com/catalog/decomposeConditional.html) refactoring so that the complex conditional is encapsulated in a separate function with a good name that captures the business rule. Optionally, for simple expressions, introduce a new variable which holds the result of the complex conditional.
Helpful refactoring examples

To get a general understanding of what this code health issue looks like - and how it might be addressed - we have prepared some diffs for illustrative purposes.
SAMPLE
complex_conditional.js
function messageReceived(message, timeReceived) {
   // Ignore all messages which aren't from known customers:
   if (!message.sender &&
       customers.getId(message.name) == null) {
   // Refactoring #1: encapsulate the business rule in a
   // function. A clear name replaces the need for the comment:
   if (!knownCustomer(message)) {
     log('spam received -- ignoring');
     return;
   }
  // Provide an auto-reply when outside business hours:
  if ((timeReceived.getHours() > 17) ||
      (timeReceived.getHours() < 8)) {
  // Refactoring #2: encapsulate the business rule.
  // Again, note how a clear function name replaces the
  // need for a code comment:
  if (outsideBusinessHours(timeReceived)) {
    return autoReplyTo(message);
  }
  pingAgentFor(message);
}
function outsideBusinessHours(timeReceived) {
  // Refactoring #3: replace magic numbers with
  // symbols that communicate with the code reader:
  const closingHour = 17;
  const openingHour = 8;
  const hours = timeReceived.getHours();
  // Refactoring #4: simple conditional rules can
  // be further clarified by introducing a variable:
  const afterClosing = hours > closingHour;
  const beforeOpening = hours < openingHour;
  // Yeah -- look how clear the business rule is now!
  return afterClosing || beforeOpening;
}
```

**Comment by leynos** [comment]
@coderabbitai review

File: weaverd/server.py [hunk 1]
```diff
@@ -41,12 +41,56 @@ def list_diagnostics(
...
+
+class ReferenceLookupError(RuntimeError):
+    """Raised when looking up symbol references fails."""
+
+    def __init__(self, exc: Exception) -> None:
+        super().__init__(f"Reference lookup failed: {exc}")
+
+
+class InvalidPositionError(ValueError):
+    """Raised when a line or character index is negative."""
```
[the remark this answers is not in the document]
Comment (leynos) [reply weaverd/server.py:80]: Has this now been resolved?

File: weaver/client.py [hunk 2]
```diff
@@ -103,11 +110,15 @@ async def rpc_call(
     stdout = typ.cast(typ.TextIO, sys.stdout if stdout is None else stdout)  # noqa: TC006
     try:
         await ensure_daemon_running(path)
-    except Exception as exc:
+    except DaemonStartError as exc:
```
[the remark this answers is not in the document]
Comment (leynos) [reply weaver/client.py:113]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

Complex Method
weaver/client.py: rpc_call

What lead to degradation?

rpc_call has a cyclomatic complexity of 9, threshold = 9
Why does this problem occur?

A Complex Method has a high cyclomatic complexity. The recommended threshold for the Python language is a cyclomatic complexity lower than 9.
How to fix it?

There are many reasons for Complex Method. Sometimes, another design approach is beneficial such as a) modeling state using an explicit state machine rather than conditionals, or b) using table lookup rather than long chains of logic. In other scenarios, the function can be split using [EXTRACT FUNCTION](https://refactoring.com/catalog/extractFunction.html). Just make sure you extract natural and cohesive functions. Complex Methods can also be addressed by identifying complex conditional expressions and then using the [DECOMPOSE CONDITIONAL](https://refactoring.com/catalog/decomposeConditional.html) refactoring.
Helpful refactoring examples

To get a general understanding of what this code health issue looks like - and how it might be addressed - we have prepared some diffs for illustrative purposes.
SAMPLE
complex_method.js
function postItem(item) {
  if (!item.id) {
    if (item.x != null && item.y != null) {
      post(item);
    } else {
      throw Error("Item must have x and y");
    }
    // extract a separate function for creating new item
    postNew(item);
  } else {
    if (item.x < 10 && item.y > 25) {
      put(item);
    } else {
      throw Error("Item must have an x and y value between 10 and 25");
    }
    // and one for updating existing items
    updateItem(item);
  }
}
function postNew(item) {
  validateNew(item);
  post(item);
}
function updateItem(item) {
  validateUpdate(item);
  put(item);
}

**Comment by leynos** [comment]
@coderabbitai please create new GitHub issues for the following:


    weaverd/unittests/test_onboard.py (1)

        62-66: Test-only custom exception is fine; consider hoisting to module scope for reuse.

        Defining OnboardToolError inline works and aligns with TRY’s guidance to avoid generic exceptions. If you expect to reuse it across multiple tests, define it once at module scope to avoid repeated class creation and to improve readability.

    weaverd/unittests/test_rpc.py (1)

        108-112: LGTM; using a test-specific exception aligns with TRY. Optional: define once at module scope.

        BoomError as a local subclass is fine here. If you end up reusing it, consider hoisting it to module scope to avoid redefinition overhead in generators.

    weaver/unittests/test_cli.py (1)

        51-55: De-duplicate FakeRunError by defining it once at module scope.

        Both tests define identical local exception classes. Define FakeRunError once at the top of the file and reuse it to reduce duplication and tiny runtime overhead.

        Apply these diffs within the selected ranges:

        -        class FakeRunError(RuntimeError):
        -            """Test-only error to simulate RPC failure."""
        -
        -        raise FakeRunError("boom")
        +        raise FakeRunError("boom")

        -        class FakeRunError(RuntimeError):
        -            """Test-only error to simulate RPC failure."""
        -
        -        raise FakeRunError("boom")
        +        raise FakeRunError("boom")

        Add this once near the imports (outside the selected ranges):

        class FakeRunError(RuntimeError):
            """Test-only error to simulate RPC failure."""

        Also applies to: 169-173

    features/conftest.py (1)

        26-34: Good: raises explicit DaemonNotReadyError and cleans up the writer. Optional: simplify timeout with anyio.fail_after.

        The polling loop and resource cleanup look correct. For clarity, consider anyio.fail_after to express the timeout and centralize the final error path.

        Example refactor:

        async def _wait_for_socket(path: Path, timeout: float = 5.0) -> None:
            try:
                with anyio.fail_after(timeout):
                    while True:
                        try:
                            _, writer = await asyncio.open_unix_connection(str(path))
                        except OSError:
                            await anyio.sleep(0.05)
                        else:
                            writer.close()
                            with contextlib.suppress(Exception):
                                await writer.wait_closed()
                            return
            except TimeoutError:
                raise DaemonNotReadyError(path, timeout) from None

    weaver/client.py (1)

        60-70: Expose retry policy as constants/parameters for maintainability.

        The fixed 50×0.1s retry policy is fine, but consider module-level constants or parameters to make behavior easy to tune and to align with test-time DaemonNotReadyError semantics.

        Possible approach (outside selected lines):

        STARTUP_RETRIES = 50
        STARTUP_SLEEP_SECS = 0.1

        Then:

        for _ in range(STARTUP_RETRIES):
            if await can_connect(socket_path):
                return
            await anyio.sleep(STARTUP_SLEEP_SECS)

    pyproject.toml (1)

        25-33: Remove duplicated dev dependency definitions

        The verification script confirms that the dev lists under [project.optional-dependencies] and [dependency-groups] are currently identical (no only-in-extras or only-in-groups) — the duplication persists with no drift yet, but poses a risk of divergence down the line.

        To maintain a single source of truth, pick one of the following:

        Option A (recommended):
        • Keep only the [dependency-groups].dev block for UV-driven workflows and drop the extras if you don’t publish a “dev” extra.

        Option B:
        • Keep only [project.optional-dependencies].dev so that pip install .[dev] continues to work, and invoke UV with --extra dev (supported by UV).

        Locations to update:

            pyproject.toml, lines 25–33 (extras)
            pyproject.toml, lines 34–41 (dependency-groups)

        To remove the duplicate dependency-group definition, apply:

        -[dependency-groups]
        -dev = [
        -    "pytest",
        -    "ruff==0.12.7",
        -    "pyright",
        -    "pytest-timeout",
        -    "pytest-bdd>=8.1.0",
        -]

    weaverd/serena_tools.py (3)

        51-58: Use a more semantically accurate base for PromptFactoryError.

        The message covers “not found or not a type”. If the attribute is missing it’s an import/attribute problem; if present but not a type it’s a type problem. If you prefer a single exception, making it an ImportError subclass better reflects the lookup failure scenario and is still explicit.

        Minimal change:

        -class PromptFactoryError(TypeError):
        +class PromptFactoryError(ImportError):

        69-74: Type error message could include the actual type for faster diagnosis.

        Including the received type makes debugging easier.

        -class ToolAttrTypeError(TypeError):
        +class ToolAttrTypeError(TypeError):
        @@
        -    def __init__(self) -> None:
        -        super().__init__("tool_attr must be SerenaTool or str")
        +    def __init__(self, received: object | None = None) -> None:
        +        got = type(received).__name__ if received is not None else "unknown"
        +        super().__init__(f"tool_attr must be SerenaTool or str (got {got})")

        And update the raise site to pass the value:

        -    if not isinstance(tool_attr, str):
        -        raise ToolAttrTypeError()
        +    if not isinstance(tool_attr, str):
        +        raise ToolAttrTypeError(tool_attr)

        76-83: UnknownSerenaToolError is helpful; consider LookupError as a base.

        This is a missing/invalid choice rather than a runtime failure; inheriting from LookupError or ValueError better matches Python conventions while keeping your custom type.

        -class UnknownSerenaToolError(RuntimeError):
        +class UnknownSerenaToolError(LookupError):

    weaverd/unittests/test_serena_tools.py (3)

        41-49: Prefer constructing ModuleNotFoundError with a name instead of mutating.

        Using the built-in constructor mirrors real-world exceptions and keeps type-checkers happier.

        -def test_is_serena_related_error_by_module_name() -> None:
        -    exc = ModuleNotFoundError()
        +def test_is_serena_related_error_by_module_name() -> None:
        +    exc = ModuleNotFoundError("msg", name="serena.fake")
             assert _is_serena_related_error("serena.fake", exc) is True

        51-54: Same here: construct with name=... rather than setting attribute.

        -def test_is_serena_related_error_by_exc_name() -> None:
        -    exc = ModuleNotFoundError()
        -    exc.name = "serena.fake"
        +def test_is_serena_related_error_by_exc_name() -> None:
        +    exc = ModuleNotFoundError("msg", name="serena.fake")
             assert _is_serena_related_error("other", exc) is True

        51-55: And for the negative case, construct with non-Serena name.

        -def test_is_serena_related_error_false() -> None:
        -    exc = ModuleNotFoundError()
        -    exc.name = "other"
        +def test_is_serena_related_error_false() -> None:
        +    exc = ModuleNotFoundError("msg", name="other")
             assert _is_serena_related_error("random", exc) is False

    features/steps/helpers.py (1)

        5-7: Unify TYPE_CHECKING usage with the project’s import alias conventions.

        You import typing as typ and also from typing import TYPE_CHECKING (with a noqa). Prefer typ.TYPE_CHECKING and drop the direct import to satisfy ICN and reduce exceptions.

        -import typing as typ
        -from typing import TYPE_CHECKING  # noqa: ICN003
        +import typing as typ
        @@
        -if TYPE_CHECKING:
        +if typ.TYPE_CHECKING:

    features/steps/test_onboard_project.py (1)

        71-75: Nested test-only exception is fine; minor polish optional.

        Defining ToolApplyError at function scope (instead of within apply) can make tracebacks slightly clearer, but this is purely a style nit for tests.

        -        class FailingTool:
        -            def apply(self) -> str:  # pragma: no cover - stub
        -                class ToolApplyError(RuntimeError):
        -                    """Test-only error to simulate tool failure."""
        -                raise ToolApplyError("boom")
        +        class ToolApplyError(RuntimeError):
        +            """Test-only error to simulate tool failure."""
        +
        +        class FailingTool:
        +            def apply(self) -> str:  # pragma: no cover - stub
        +                raise ToolApplyError("boom")

    weaverd/unittests/test_definition.py (2)

        76-91: Localize StopAsyncIteration handling and remove unreachable raise.

        The helper works, but the broad try/except slightly obscures intent and the raise after pytest.fail is effectively unreachable on early termination. Consider tightening the scope to the collection loop for clarity.

        Apply this refactor:

         async def _collect_symbols_from_results(
             results: cabc.AsyncIterator[Symbol], expected_count: int
         ) -> list[Symbol]:
             """Helper to collect symbols from async iterator and verify count."""
        -    symbols: list[Symbol] = []
        -    try:
        -        for _ in range(expected_count):
        -            symbols.append(await anext(results))
        -        # Verify no more symbols remain
        -        with pytest.raises(StopAsyncIteration):
        -            await anext(results)
        -    except StopAsyncIteration:
        -        if len(symbols) != expected_count:
        -            pytest.fail(f"Expected {expected_count} symbols, got {len(symbols)}")
        -        raise
        -    return symbols
        +    symbols: list[Symbol] = []
        +    for i in range(expected_count):
        +        try:
        +            symbols.append(await anext(results))
        +        except StopAsyncIteration:
        +            pytest.fail(f"Expected {expected_count} symbols, got {i}")
        +    # Verify no more symbols remain
        +    with pytest.raises(StopAsyncIteration):
        +        await anext(results)
        +    return symbols

        181-181: Minor: prefer capturing the exception to assert the message explicitly.

        Regex match is fine; alternatively, capture and assert to avoid regex edge cases:

        with pytest.raises(SerenaAgentNotFoundError) as ei:
            await anext(server.handle_get_definition("foo.py", 1, 0))
        assert str(ei.value) == "serena-agent not found"

    weaverd/unittests/test_references.py (2)

        65-65: Minor: assert on captured exception message instead of regex.

        Regex is acceptable; capturing avoids escaping pitfalls and clarifies intent:

        with pytest.raises(SerenaAgentNotFoundError) as ei:
            await builtins.anext(server.handle_list_references("foo.py", 1, 0))
        assert str(ei.value) == "serena-agent not found"

        73-76: Nested test-only exception is fine; consider module-scope for readability.

        Defining ListReferencesToolError inside the method works and keeps scope tight. If you prefer slightly clearer traces and reuse, lift it to module scope and just raise it here. Optional change; current approach is acceptable.

File: weaver/client.py [hunk 3]
```diff
@@ -92,22 +99,35 @@ async def _stream_response(reader: asyncio.StreamReader, stdout: typ.TextIO) ->
...
 
-    reader, writer = await asyncio.open_unix_connection(str(path))
+async def _execute_rpc_request(
+    reader: asyncio.StreamReader,
+    writer: asyncio.StreamWriter,
+    method: str,
+    params: JSONObject | None,
+    stdout: typ.TextIO,
+) -> bool:
+    """Send an RPC request and stream the response."""
```
Comment (leynos) [inline weaver/client.py]: @coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix.

**Comment by leynos** [comment]
@coderabbitai pause
